### PR TITLE
[WIP] Use OJDK code to resolve OJDK MH CP Entry

### PR DESCRIPTION
**Pull request by babsingh**
In OJDK, [MHN.linkMethodHandleConstant](https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/09e79a7ee9dce92e1741867f43e9b20c58e16f27/src/java.base/share/classes/java/lang/invoke/MethodHandleNatives.java#L602-L619) resolves a MethodHandle constant
pool entry. It is being incorporated in OpenJ9 to be used with OJDK MHs.
It has certain fast-paths which improve the performance.

CI builds are skipped because OJDK MHs are currently disabled in OpenJ9.

[ci skip]

Co-authored-by: Jack Lu <Jack.S.Lu@ibm.com>
Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
fyi @DanHeidinga

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java [hunk 1]
```diff
@@ -369,16 +369,34 @@ private static final MethodHandle sendResolveMethodHandle(
...
+		case 5: /* invokeVirtual */
+		case 6: /* invokeStatic */
+		case 7: /* invokeSpecial */
+		case 8: /* newInvokeSpecial */
+		case 9: /* invokeInterface */
+			type = MethodTypeHelper.vmResolveFromMethodDescriptorString(typeDescriptor, loader, null);
+			break;
+		default:
+			/* Can never happen */
+			throw new UnsupportedOperationException();
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java:390]: Throw an InternalError here with details of the problem

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java [hunk 2]
```diff
@@ -369,16 +369,34 @@ private static final MethodHandle sendResolveMethodHandle(
 			String name,
 			String typeDescriptor,
 			ClassLoader loader) throws Throwable {
-		try {
+		MethodType type = null;
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java:372]: Why was this moved out of the try block?  Does the OJDK code not need the IllegalAccessException -> IllegalAccessError mapping?

**Review by DanHeidinga — COMMENTED** [review]
This removes the references to comments about `accessCheckArgRetTypes` - have you confirmed we don't need to do the access check?

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java (hunk 2 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java:372]: > Why was this moved out of the try block? 

It is used by both the implementations. By moving it out, it is shared by both the implementations.

> Does the OJDK code not need the IllegalAccessException -> IllegalAccessError mapping?

Yes, it does need it. The mapping is handled in `linkMethodHandleConstant`, which is OJDK code. So, we don't need to handle the mapping in OpenJ9 code. `linkMethodHandleConstant` invokes `mapLookupExceptionToError` to handle the mapping.

`mapLookupExceptionToError`: https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/ac04db2e39441f285acebb657aa388043a894770/src/java.base/share/classes/java/lang/invoke/MethodHandleNatives.java#L621-L641

**Comment by babsingh** [comment]
> This removes the references to comments about accessCheckArgRetTypes - have you confirmed we don't need to do the access check?

`accessCheckArgRetTypes` throws an `IllegalAccessException: Class '{0}' no access to: class '{1}'`.

`mapLookupExceptionToError`  converts an `IllegalAccessException` to an `IllegalAccessError`. So, I believe that `linkMethodHandleConstant` should also throw a similar `IllegalAccessException` as `accessCheckArgRetTypes`. Let me confirm with an example.

**Comment by babsingh** [comment]
> This removes the references to comments about accessCheckArgRetTypes - have you confirmed we don't need to do the access check?

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java (hunk 1 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleResolver.java]: Completed.

**Comment by DanHeidinga** [comment]
Those exceptions get thrown by the native code in OJDK.  We implemented those checks in Java instead using `accessCheckArgRetTypes`.  Where in our implementation of the natives is the equivalent checks?  I expect it's in one of the other PRs but having trouble finding it....

**Comment by babsingh** [comment]
> Where in our implementation of the natives is the equivalent checks? I expect it's in one of the other PRs but having trouble finding it....

It should be the `MethodHandleNatives` PR: https://github.com/eclipse/openj9/pull/10690. `MemberName` resolution should fail in `MethodHandleNatives_resolve`.

**Comment by babsingh** [comment]
> Those exceptions get thrown by the native code in OJDK

The exceptions are not thrown by the native code in OJDK. The code references in https://github.com/eclipse/openj9/pull/11371#issuecomment-740700038 points to the Java code in OJDK. The exceptions are generated and thrown by the Java code in OJDK. OpenJ9 should also consume this code. But, the indicator to execute this Java code is the failure of `MemberName` resolution, which partially resides with `MethodHandleNatives_resolve`. There is also OJDK Java code involved with `MemberName` resolution, which can lead to failure.

**Comment by DanHeidinga** [comment]
@babsingh and @fengxue-IS I've been digging through this code and don't see the equivalent of `accessCheckArgRetTypes` in our code, including PR: #10690.

Can you point me at the code that walks through the classes in the MethodType and ensures they are accessible from the `lookup()`'s `accessClass`?

**Comment by DanHeidinga** [comment]
Jenkins test sanity zlinux jdk11

**Comment by DanHeidinga** [comment]
Running builds on this in the mean time as this may be ready to merge depending on the answers to questions above

**Comment by DanHeidinga** [comment]
Reminder to self - Next steps for this PR depend on answer to questions in: https://github.com/eclipse/openj9/pull/11371#issuecomment-745759137

**Comment by babsingh** [comment]
> Can you point me at the code that walks through the classes in the MethodType and ensures they are accessible from the lookup()'s accessClass?

Jack and I couldn't find the equivalent access checks in the RI. Even, our `MHN.resolve` doesn't do those access checks. Jack is writing a `ldc` test case with inaccessible `MT.ptyes` to see how those access checks are enforced in RI. Updates will be posted soon.

**Comment by DanHeidinga** [comment]
> Jack is writing a `ldc` test case with inaccessible `MT.ptyes` to see how those access checks are enforced in RI. Updates will be posted soon.

Any progress on this?  I'm about to start vacation and am unlikely to look in on this PR over the holidays

**Comment by babsingh** [comment]
> Any progress on this?

I found the following links: [Lookup.checkAccess](https://github.com/openjdk/jdk/blob/12297a00eff8756a2f09007d8e79bfb71f3b26b5/src/java.base/share/classes/java/lang/invoke/MethodHandles.java#L3821) and [VerifyAccess.isMemberAccessible](https://github.com/openjdk/jdk/blob/12297a00eff8756a2f09007d8e79bfb71f3b26b5/src/java.base/share/classes/sun/invoke/util/VerifyAccess.java#L50-L96).

Instead of doing the access check on the `ptypes` and `rtype` of a `MethodType`, RI does an access check on the resolved `MemberName` (`VerifyAccess.isMemberAccessible`). These checks are done in Java, and the corresponding call stack looks like:
```
- MethodHandleNatives.linkMethodHandleConstant <<<< Called from MethodHandleResolver.sendResolveMethodHandle
-- Lookup.linkMethodHandleConstant
--- Lookup.getDirectMethodForConstant
---- ... [2-3 different calls/paths]
----- [Lookup.checkMethod || Look.checkField]
------ Lookup.checkAccess <<<< performs the access check
------- VerifyAccess.isMemberAccessible <<<< evaluates the JVM linkage rules for access to the given method on behalf of a caller class
```

Jack tried to write a test case with generated bytecodes where `ldc` was invoked on a `MethodType` with inaccessible `ptypes` and `rtype`. He has me informed that a `MethodType` with inaccessible `ptypes` and `rtype` cannot be created.

Due to lack of a test case, I don't have concrete evidence that `Lookup.accessCheckArgRetTypes == VerifyAccess.isMemberAccessible`.

In https://github.com/eclipse/openj9/issues/10913, I have added a note that any missing `IllegalAccessError`s during functional testing may be due to `Lookup.accessCheckArgRetTypes != VerifyAccess.isMemberAccessible`. If such a case arises, then appropriate action will be taken.

**Comment by DanHeidinga** [comment]
Check the IBM internal repos and use the `history` branch there to look at the `blame` for `Lookup.accessCheckArgRetTypes`.  That should hopefully point you at the test cases which prompted the function to be added

**Comment by babsingh** [comment]
> Check the IBM internal repos and use the `history` branch there to look at the `blame` for `Lookup.accessCheckArgRetTypes`. That should hopefully point you at the test cases which prompted the function to be added

These changes were done in the external OpenJ9 repo after open-sourcing: https://github.com/eclipse/openj9/pull/2675.

There is a mention of a `failing test suite` in https://github.com/eclipse/openj9/pull/2675#issuecomment-415421049. But, no name is specified for the `failing test suite`.

The commit message states `Throw IllegalAccessError if we call do method resolution referring to inaccessible types`.

This somewhat matches the behaviour of `VerifyAccess.isMemberAccessible` as described here: https://github.com/openjdk/jdk/blob/12297a00eff8756a2f09007d8e79bfb71f3b26b5/src/java.base/share/classes/sun/invoke/util/VerifyAccess.java#L63-L86.

**Comment by babsingh** [comment]
@DanHeidinga This PR's description has been changed so that it can be merged. The concerns about access checking are being tracked in https://github.com/eclipse/openj9/issues/10913. Once a fix for access checks is identified, it will be merged in a separate PR.

**Comment by DanHeidinga** [comment]
@babsingh it's not the description that's blocking the merge of this.  Right now there are clear indicators in the code of a serious potential issue - the fact that we don't know if the class access checks are required in the OJDK impl or handled in another way or simply missing due to the way we're using their code - and merging this removes those indicators.

I'm not comfortable with merging this until we have an answer on those access checks

**Comment by DanHeidinga** [comment]
> I'm not comfortable with merging this until we have an answer on those access checks

We've been pushing this question forward since at least Oct of last year.  Let's get it resolved so we can merge this.

**Comment by babsingh** [comment]
> answer on those access checks

The micro-tests for access checks are failing due to other functional issues, which are related to error checking + segfaults. This is blocking us to verify those access checks. First, the other issues will need to be resolved. Only then, the concerns about access checks can be addressed. It will be an incremental process, and it will depend on the number of other functional issues encountered before the access checks can be verified. I am not sure how much time this would take.

Jack, Liqun, Nazim and I have been working on resolving the functional issues. Some of them are documented in https://github.com/eclipse/openj9/issues/11458.

I have re-added the indicator for the `access check` in the new code. It is situated in the same place as the old code after the MethodType instance is generated. The only difference is that OJDK code is used to resolve MH CP entries.

I am happy to wait until the access checks are verified. At the same time, I am worried that the amount of code/PRs to be reviewed will pile up by the time access checks are verified.

Changing the PR to WIP until access checks are verified.

**Comment by DanHeidinga** [comment]
@babsingh Did you determine these checks were required (or not)?  Updating this with a status would make it clearer why the PR was closed

**Comment by babsingh** [comment]
> Did you determine these checks were required (or not)? Updating this with a status would make it clearer why the PR was closed

I was cleaning up my branches in my remote fork, and I accidentally deleted `ojdk_jsr292_v3`. I have recreated the deleted branch. But, I don't have permissions to reopen this PR. @DanHeidinga are you able to reopen this PR?

**Comment by DanHeidinga** [comment]
The button is greyed out with the tooltip: "The ojdk_jsr292_v3 branch was force-pushed or recreated"

I don't know how to resolve that short of opening a new PR

**Comment by babsingh** [comment]
Created https://github.com/eclipse/openj9/pull/11827 to replicate this PR.
